### PR TITLE
Fix `accumulate_grad_batches` on init

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Just a small fix to make sure that accumulate_grad_batches is an int all the time and we use gradient accumulation scheduler to define it's value.

```py
self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(0)
```

the reason to have this one even here though it's updated inside on_train_epoch_start is because I have seen users using it to calculate total training_steps which is usually used while defining some of the schedulers. So this is required at this point.

also in case a scheduler is passed inside callbacks then that is also taken care off now.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/training_trick_connector.py [hunk 1]
```diff
@@ -55,4 +54,8 @@ def configure_accumulated_gradients(self, accumulate_grad_batches: Union[int, Di
             schedule = {0: accumulate_grad_batches}
             self.trainer.accumulation_scheduler = GradientAccumulationScheduler(schedule)
         else:
-            raise TypeError("Gradient accumulation supports only int and dict types")
+            raise MisconfigurationException(
+                f"Gradient accumulation supports only int and dict types, got {accumulate_grad_batches}"
+            )
+
+        self.trainer.accumulate_grad_batches = self.trainer.accumulation_scheduler.get_accumulate_grad_batches(0)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/training_trick_connector.py:61]: I think what you are fixing here is this line, where you are already setting the value here instead of waiting for the `GradientAccumulationScheduler` to do it in the `on_train_epoch_start` hook. 
A bit more context on the PR description would help the reviewer next time :)

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/training_trick_connector.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/training_trick_connector.py:61]: is `trainer.accumulate_grad_batches` always expected to be an int? could we add a typehint for this if so?

tangentially related: what happens if someone passes both `accumulate_grad_batches=X` and a `GradientAccumulationScheduler` callback to the trainer? Should we detect this and raise a misconfiguration error in this case?

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/training_trick_connector.py:61]: @awaelchli updated the description.
thanks @ananthsub will check this :)

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -186,11 +186,41 @@ def test_trainer_accumulate_grad_batches_zero_grad(tmpdir, accumulate_grad_batch
...
         trainer.fit(model)
 
         assert sgd_zero_grad.call_count == math.ceil(trainer.limit_train_batches / accumulate_grad_batches)
 
 
+@pytest.mark.parametrize(
+    ["accumulate_grad_batches", "expected_call_count"],
+    [
+        ({1: 2, 3: 4}, 23),  # 10+5+5+3
+        ({0: 2, 2: 1}, 30),  # 5+5+10+10
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:199]:
```suggestion
        ({1: 2, 3: 4}, 10 + 5 + 5 + 3),
        ({0: 2, 2: 1}, 5 + 5 + 10 + 10),
```

**Review by carmocca — APPROVED** [review]

**Comment by rohitgr7** [comment]
moved the logic to callback connector since `GradientAccumulateCallback` is also a callback and we need to take care for that if that is passed by user inside callback. if a callback is passed we check if the other argument is passed or not.. if both are present then a minconfig error is raised (we can check it to warning if required) else callback is considered as well.

**Comment by rohitgr7** [comment]
seems like there is some issue with ipus + accumulate_grad_batches.. need to explore more on this.

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 3]
```diff
@@ -185,27 +183,21 @@ def _convert_to_poptorch_loader(
...
+                "IPUs currently does not support different `accumulate_grad_batches` at different epoch."
             )
-        # save the original value which will be used to update the global step progress
-        self._original_accumulate_grad_batches = accumulate_grad_batches
-        if accumulate_grad_batches > 1:
-            # TODO(@tchaton): Add support for accumulate_grad_batches being a dictionary
-            self.lightning_module.trainer.accumulation_scheduler = GradientAccumulationScheduler({0: 1})
+
+        # TODO(@tchaton): Add support for accumulate_grad_batches being a dictionary
+        accumulation_scheduler.scheduling.update({0: 1})
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/ipu.py:200]: I believe, dict can be supported but if we have multiple accumulation factor at different epoch.. then it's a problem.. so updated it accordingly along with the error message.

File: tests/accelerators/test_ipu.py [hunk 4]
```diff
@@ -310,9 +310,9 @@ def on_predict_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, da
 @RunIf(ipu=True)
 def test_accumulate_grad_batches_dict_fails(tmpdir):
     model = IPUModel()
-    trainer = Trainer(default_root_dir=tmpdir, ipus=1, accumulate_grad_batches={0: 1})
+    trainer = Trainer(default_root_dir=tmpdir, ipus=1, accumulate_grad_batches={1: 2})
```
Comment (rohitgr7) [inline tests/accelerators/test_ipu.py:313]: also updated the test. {0: 1} is fine but {1:2} is not

**Comment by rohitgr7** [comment]
> Build Error! No Linked Issue found. Please link an issue or mention it in the body using #<issue_id>

apologies for this.. initially I thought this is going to be a simple fix.. but turns out there were 2 more issues.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 5]
```diff
@@ -87,10 +89,40 @@ def on_trainer_init(
...
         # it is important that these are the last callbacks to run
         self.trainer.callbacks = self._reorder_callbacks(self.trainer.callbacks)
 
+    def _configure_accumulated_gradients(self, accumulate_grad_batches: Union[int, Dict[int, int]]) -> None:
+        grad_accum_callback = [cb for cb in self.trainer.callbacks if isinstance(cb, GradientAccumulationScheduler)]
+
+        if grad_accum_callback:
+            if accumulate_grad_batches is not None:
+                raise MisconfigurationException(
+                    "You have set both `accumulate_grad_batches` and passed an "
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:105]: let's mention the action what the user needs to do. remove either the trainer argument or remove the callback from the callbacks list.

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 6]
```diff
@@ -87,10 +89,40 @@ def on_trainer_init(
...
+            if isinstance(accumulate_grad_batches, dict):
+                grad_accum_callback = GradientAccumulationScheduler(accumulate_grad_batches)
+            elif isinstance(accumulate_grad_batches, int):
+                grad_accum_callback = GradientAccumulationScheduler({0: accumulate_grad_batches})
+            else:
+                raise MisconfigurationException(
+                    f"Gradient accumulation supports only int and dict types, got {accumulate_grad_batches}"
+                )
+
+        self.trainer.callbacks.append(grad_accum_callback)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:122]: if the callback was already there, we append it again. this should be avoided

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 7]
```diff
@@ -87,10 +89,40 @@ def on_trainer_init(
...
+            if accumulate_grad_batches is None:
+                accumulate_grad_batches = 1
+
+            if isinstance(accumulate_grad_batches, dict):
+                grad_accum_callback = GradientAccumulationScheduler(accumulate_grad_batches)
+            elif isinstance(accumulate_grad_batches, int):
+                grad_accum_callback = GradientAccumulationScheduler({0: accumulate_grad_batches})
+            else:
+                raise MisconfigurationException(
+                    f"Gradient accumulation supports only int and dict types, got {accumulate_grad_batches}"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:119]: for consistency, this error message should end with a period like all the other messges

```suggestion
                    f"Gradient accumulation supports only int and dict types, got {accumulate_grad_batches}."
```

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -140,7 +140,7 @@ def __init__(
         track_grad_norm: Union[int, float, str] = -1,
         check_val_every_n_epoch: int = 1,
         fast_dev_run: Union[int, bool] = False,
-        accumulate_grad_batches: Union[int, Dict[int, int]] = 1,
+        accumulate_grad_batches: Optional[Union[int, Dict[int, int]]] = None,
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:143]: + we probably want to mention this change of default value explicitly in the changelog

File: pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -435,8 +435,6 @@ def __init__(
         # default .predict() loop
         self.predict_loop = PredictionLoop()
 
-        self.weights_summary = weights_summary
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/trainer.py:438]: no, it doesn't. just a minor update. this is assigned to the trainer where it should be.

File: pytorch_lightning/trainer/connectors/callback_connector.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/callback_connector.py]: good catch!

**Comment by Borda** [comment]
> > Build Error! No Linked Issue found. Please link an issue or mention it in the body using #<issue_id>
> 
> apologies for this.. initially I thought this is going to be a simple fix.. but turns out there were 2 more issues.

That is fine, take it easy... 
@tchaton the bot is quite annoying 🐰

File: pytorch_lightning/trainer/trainer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:438]: I disagree. Now the connector is setting `self.trainer.weights_summary = ...` A pattern we want avoid and eventually remove and refactor.
